### ISSUE TITLE
Screensharing freezes with "[WARN] - pipewire: out of buffers"

### ISSUE THREAD

**Issue by David96** [emersion/xdg-desktop-portal-wlr#69]
Hey everyone,

after some time, screen sharing using Firefox, e.g. on https://mozilla.github.io/webrtc-landing/gum_test.html or BBB, freezes. The easiest and fastest way to reproduces this seems to be to draw with my stylus in xournalpp but I also had this happening in a session without using the pen (after a considerably longer period of time).  I attached a xdpw debug log - the relevant part should be this line: "[WARN] - pipewire: out of buffers". Restarting the screen share in Firefox fixes the issue (edit: until it occurs again that is, so not quite fixed, just temporarily. I chose a bad wording there).
I'm running sway from master branch and xdpw with the screen selector pull request (to be easily able to share the internal screen which I'm able to draw on with my pen).

[debug.log](https://github.com/emersion/xdg-desktop-portal-wlr/files/5644634/debug.log)

Pipewire version: 0.3.17-1
libpipewire02: 0.2.7-1 (I think FF uses this?)
firefox-wayland-hg: r625715.0a7de2b53f8a-1 (but also happened in Chromium, so I don't think FF is to blame here)

Let me know if you need any other information.
Thanks a lot for your work on this!

**Comment by danshick** [comment]
Hmmm, I've never seen or heard of this kind of issue before. Maybe a recent update in a dependency (that I haven't gotten yet) is to blame? Or maybe something unique to your setup. I know where in xdpw that log message is from, but I'm not sure exactly why it should be happening.


When I get a chance, I'll compare pipewire versions with you and see if it's misbehaving.

**Comment by David96** [comment]
That's what I meant with "with the screen selector pull request", should have made that clearer. Somehow I missed the -o option of xdpw :facepalm: Anyway, I was able to reproduce the issue also without that pull request. Same debug log, with the same infos before the warning of the state changing to connecting to paused to streaming and then running out of buffers.

I'm actually experiencing this since a few weeks (and that was actually also the first time I had to use screen sharing a bit more) so it's hard to pinpoint any specific update as the cause here.

If you have any wild guesses what could cause or influence this behaviour I'm of course happy to do some investigation in that direction.

Regarding my setup: I'm using Arch with testing repos enabled and have a AMD Ryzen 7 3700U CPU with Radeon Vega Mobile Gfx.

**Comment by danshick** [comment]
Hmmm, okay. I should have read your whole issue a little better, that's on me. I haven't had a chance to try to reproduce yet, but if it is related to your card, I can't help as I only have integrated intel graphics everywhere.

One thing that might be helpful are pipewire logs. If you manually stop pipewire with `systemctl -u stop pipewire.service` and `systemctl -u stop pipewire.socket`, you can run it manually from a command line. If you prefix it with the correct environment variable you can increase the verbosity progressively. Usually level 3 is enough. Then you can just capture the outputs. Something like:

```sh
PIPEWIRE_DEBUG=3 pipewire 2>&1 | tee pipewire.log
```

**Comment by David96** [comment]
Hmm, this is kinda weird to debug. I tried reproducing it while creating a log and wasn't quite able to, even after ~30min nothing happened (usually it takes about a minute or so). Not sure, but watching Netflix (using FF in this case) seems to more or less help triggering the problem, after playing around a bit, starting pipewire via systemd a few times, opening Netflix in a popup window I was able to also reproduce it while creating a debug log. But as I see it, it shows nothing after the connection to xdpw was established 
[pipewire.log](https://github.com/emersion/xdg-desktop-portal-wlr/files/5648871/pipewire.log)

Do you know in which memory the buffers that apparently run out lie? Ie graphics memory or main one? In case of gfx, maybe there's some leak? Main memory I should have enough that I shouldn't run out of that (constantly around 30% of 16GB).

**Comment by danshick** [comment]
> Do you know in which memory the buffers that apparently run out lie? Ie graphics memory or main one? In case of gfx, maybe there's some leak? Main memory I should have enough that I shouldn't run out of that (constantly around 30% of 16GB).

The buffers pipewire would be referring to (as of today) would have go be in main memory, not gfx memory.

Lemme look at these logs and see id anything jumps out.

**Comment by danshick** [comment]
```
[I][000003701.818778][module-protocol-native.c:279 handle_client_error()] protocol-native 0x5583e6cbe690: client 0x5583e6d516b0 disconnected
```
https://github.com/PipeWire/pipewire/blob/6566f52a7ef09bda6a3ff684d608951851315544/src/modules/module-protocol-native.c#L279

I think pipewire appears to think a client has disconnected which kicks off the node destruction and leads to the out of buffers error. It's hard for me to tell from the addresses, but I think that client is Firefox and not us.

You mentioned that this happens in both Firefox and Chromium, but they actually share the same WebRTC implementation. Can you try another pipewire client like the rest Python script or obs-xdg-portal and let me know if the same issue occurs?

We could also get some webrtc logs from Firefox.

**Comment by David96** [comment]
From a quick test I couldn't reproduce the issue using the python script but I didn't have the time to do a very extensive one. Furthermore for some reason FF just doesn't output any log data for me, whether I use the environment variable or set a logfile via about:networking. Will investigate further but that'll take some more days/weeks, I don't have much time currently.

But apparently you're right in that it's not the fault of xdpw.

**Comment by danshick** [comment]
> But apparently you're right in that it's not the fault of xdpw.

Strange FF won't give you logs, not sure why.

I'll close this issue then, but I'm more than happy to continue troubleshooting, so feel free to keep posting on this issue after it's closed. Or you can mention my nick (danshick) on freenode #sway.

**Comment by minhduc0711** [comment]
I also frequently run into this problem. I'm not able to find a way to reproduce it as the error just happens randomly after sharing the screen for a while.

I'm using `fedora-firefox-wayland-bin`, on Arch Linux with Intel UHD Graphics 620. I installed xdpw with `xdg-desktop-portal-wlr-git` from AUR.

Here are my logs for xdpw and Firefox, but there is nothing interesting I suppose.
[xdpw.log](https://github.com/emersion/xdg-desktop-portal-wlr/files/5713641/xdpw.log)
[firefox.log](https://github.com/emersion/xdg-desktop-portal-wlr/files/5713642/firefox.log)

**Comment by danshick** [comment]
Thanks for the logs. I'll give them a look. I share my screen for a few hours at a time at least weekly with a similar setup and I haven't been able to reproduce.

**Comment by soyuka** [comment]
Hi I had the same issue and had to update to pipewire-git, also using xdg-desktop-portal-wlr-git

**Comment by soyuka** [comment]
Nevermind, tried again yesterday with firefox-developer-edition and had a freeze after ~20 min. I do think it's related to pipewire more then the portal.

**Comment by danshick** [comment]
> Nevermind, tried again yesterday with firefox-developer-edition and had a freeze after ~20 min. I do think it's related to pipewire more then the portal. 

Hmmm, I wish there were something more obvious in the logs. It seems the common thread here is Firefox/Chromium, but I'm not sure if we can blame WebRTC. And I still can't replicate this.

**Comment by minhduc0711** [comment]
I tried leaving the screen share on for 45 minutes while I was afk, and it worked fine until I came back and opened a bunch of new Reddit tabs. I got the error after switching over to 2-3 tabs. Can you try to reproduce this on your setup?

**Comment by danshick** [comment]
> I tried leaving the screen share on for 45 minutes while I was afk, and it worked fine until I came back and opened a bunch of new Reddit tabs (about 20 tabs). I got the error after switching over to 2-3 tabs. Can you try to reproduce this on your setup?

Just to clarify, you were screen sharing to one of those tabs and opening new tabs and switching between them caused the crash? Where you casting the same display that the tabs were on?

**Comment by minhduc0711** [comment]
Yes all of these are correct.

**Comment by danshick** [comment]
#87 probably fixes this, as I believe this is the same as #81 and this issue was closed prematurely. I'll mark this issue as a duplicate, even though it precedes #81.
